libdnf/conf/vars.cpp

```diff
@@ -204,21 +205,53 @@ std::string Vars::substitute(const std::string & text) const {
     return res;
 }
 
-void Vars::load(const std::string & installroot, const std::vector<std::string> & directories) {
-    detect_vars(installroot);
+void Vars::set(const std::string & name, const std::string & value, Priority prio) {
+    auto it = variables.find(name);
+    if (it != variables.end()) {
+        if (it->second.priority > prio) {
+            return;
+        }
+        it->second.value = value;
+        it->second.priority = prio;
+        return;
+    }
+    variables.insert({name, {value, prio}});
+}
 
+void Vars::load(const std::string & installroot, const std::vector<std::string> & directories) {
     load_from_env();
 
     for (const auto & dir : directories) {
         load_from_dir(std::filesystem::path(installroot) / dir);
     }
+
+    detect_vars(installroot);
 }
 
 void Vars::detect_vars(const std::string & installroot) {
-    auto arch = detect_arch();
-    values["arch"] = arch;
-    values["basearch"] = get_base_arch(arch.c_str());
-    values["releasever"] = detect_release(installroot);
+    auto it = variables.find("arch");
+    if (it == variables.end()) {
+        variables.insert({"arch", {detect_arch(), Priority::AUTO}});
+    } else if (it->second.priority <= Priority::AUTO) {
+        it->second.value = detect_arch();
+        it->second.priority = Priority::AUTO;
+    }
+
+    it = variables.find("basearch");
+    if (it == variables.end()) {
+        variables.insert({"basearch", {get_base_arch(variables["arch"].value.c_str()), Priority::AUTO}});
+    } else if (it->second.priority <= Priority::AUTO) {
+        it->second.value = variables["arch"].value.c_str();
+        it->second.priority = Priority::AUTO;
+    }
+
+    it = variables.find("releasever");
+    if (it == variables.end()) {
+        variables.insert({"releasever", {detect_release(installroot), Priority::AUTO}});
```

Comment:
I am really sorry but I looked at `detect_release()` and I am not happy with it. It returns an empty string when detection fails. The empty string is set in vars therefore $releasever in URL is replaced by empty string. Then it reports that URL of repository is not valid, but it is difficult to detect that the problem is in detection of releasever and then in substitution. Again this not a problem of PR.